tsdb/head.go

```diff
@@ -572,32 +645,217 @@ func (h *Head) Init(minValidTime int64) error {
 		return nil
 	}
 
+	start := time.Now()
 	level.Info(h.logger).Log("msg", "Replaying WAL, this may take awhile")
+
+	level.Info(h.logger).Log("msg", "creating series from WAL")
+	var err error
+	h.refSeries, h.multiRef, err = h.loadSeriesFromWAL()
+	if err != nil {
+		h.seriesReplayCorruption = true
+		return err
+	}
+
+	// No corruption in loading series. Hence set these to nil to garbage collect it.
+	defer func() {
+		h.refSeries = nil
+		h.multiRef = nil
+	}()
+
+	level.Info(h.logger).Log("msg", "m-mapping the on-disk chunks")
+	if err := h.loadMmappedChunks(h.refSeries); err != nil {
+		level.Error(h.logger).Log("msg", "loading on-disk chunks failed", "err", err)
+		h.metrics.mmapChunkCorruptionTotal.Inc()
+		// Repair is best effort here. If it fails, data will be recovered from WAL.
+		// Hence we wont lose any data (given WAL is not corrupt).
+		h.repairMmappedChunks(err, h.refSeries)
+	}
+
+	level.Info(h.logger).Log("msg", "adding remaining samples")
+	if err := h.loadSamplesFromWAL(h.refSeries, h.multiRef); err != nil {
+		return err
+	}
+
+	level.Info(h.logger).Log("msg", "finished replaying WAL", "duration", time.Since(start).String())
+	return nil
+}
+
+// PostWALRepairRecovery is used to recover the remaining samples from WAL if the corruption
+// occured in the series loading phase. This must be called iff Head.Init encountered a WAL corruption
+// and a repair was called on the WAL.
+// TODO(codesome): Add a test for this.
```

Comment:
is this still a todo?

Reply:
Yes, I will add it eventually before this PR is merged (after we have reviewed the verified the other integrations)